common/archiver/filestore/historyArchiver.go

```diff
@@ -0,0 +1,284 @@
+		return archiver.ErrArchiveNonRetriable
+	}
+
+	historyIterator := h.historyIterator
+	if historyIterator == nil {
+		var err error
+		historyIterator, err = archiver.NewHistoryIterator(request, h.historyManager, h.historyV2Manager, h.config.HistoryIteratorConfig, nil, nil)
+		if err != nil {
+			// this should not happen
+			logger.Error(archiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errConstructHistoryIterator), tag.Error(err))
+			return archiver.ErrArchiveNonRetriable
+		}
+	}
+
+	historyBatches := []*shared.History{}
+	for historyIterator.HasNext() {
+		currHistoryBatches, err := getNextSetofBatches(ctx, historyIterator)
+		if err != nil {
+			logger.Error(archiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errReadHistory), tag.Error(err))
+			return archiver.ErrArchiveNonRetriable
+		}
+
+		if historyMutated(request, currHistoryBatches, !historyIterator.HasNext()) {
+			logger.Error(archiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errHistoryMutated))
+			return archiver.ErrArchiveNonRetriable
+		}
+
+		historyBatches = append(historyBatches, currHistoryBatches...)
+	}
+
+	encodedHistoryBatches, err := encodeHistoryBatches(historyBatches)
+	if err != nil {
+		logger.Error(archiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errEncodeHistory), tag.Error(err))
+		return archiver.ErrArchiveNonRetriable
+	}
+
+	dirPath := getDirPathFromURI(URI)
+	if err = mkdirAll(dirPath); err != nil {
+		logger.Error(archiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errMakeDirectory), tag.Error(err))
+		return archiver.ErrArchiveNonRetriable
+	}
+
+	filename := constructFilename(request.DomainID, request.WorkflowID, request.RunID, request.CloseFailoverVersion)
+	if err := writeFile(path.Join(dirPath, filename), encodedHistoryBatches); err != nil {
+		logger.Error(archiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errWriteFile), tag.Error(err))
+		return archiver.ErrArchiveNonRetriable
+	}
+
+	return nil
+}
+
+func (h *historyArchiver) Get(
+	ctx context.Context,
+	URI string,
+	request *archiver.GetHistoryRequest,
+) (*archiver.GetHistoryResponse, error) {
+	if !h.ValidateURI(URI) {
+		return nil, errors.New(errInvalidURI)
+	}
+
+	if err := validateGetRequest(request); err != nil {
+		return nil, errInvalidGetHistoryRequest
+	}
+
+	dirPath := getDirPathFromURI(URI)
+	exists, err := directoryExists(dirPath)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, errHistoryNotExist
+	}
+
+	var token *getHistoryToken
+	if request.NextPageToken != nil {
+		token, err = deserializeGetHistoryToken(request.NextPageToken)
+		if err != nil {
+			return nil, errGetHistoryTokenCorrupted
+		}
+	} else if request.CloseFailoverVersion != nil {
+		token = &getHistoryToken{
+			CloseFailoverVersion: *request.CloseFailoverVersion,
+			NextBatchIdx:         0,
+		}
+	} else {
+		filenames, err := listFilesByPrefix(dirPath, constructFilenamePrefix(request.DomainID, request.WorkflowID, request.RunID))
```

Comment:
This chunk of code feels like its at a different level of abstraction than other code in this methods. Can we factor it out into its own method?

Reply:
Created a separate method.